[SPMD] Introduce graph_optimization_pass and comm_fusion_with_cat

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98580
* #98579
* #98578
* __->__ #98285
* #98284
* #98182

This PR add `graph_optimization_pass` decorator which should be wrapped by all graph optimization passes. This PR also introduces the first graph optimization, `comm_fusion_with_cat`, as the first use case of `graph_optimization_pass`.

Differential Revision: [D44661608](https://our.internmc.facebook.com/intern/diff/D44661608/)

fegin committed:
[SPMD] Introduce graph_optimization_pass and comm_fusion_with_cat

This PR add `graph_optimization_pass` decorator which should be wrapped by all graph optimization passes. This PR also introduces the first graph optimization, `comm_fusion_with_cat`, as the first use case of `graph_optimization_pass`.

Differential Revision: [D44661608](https://our.internmc.facebook.com/intern/diff/D44661608/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+
+
+def graph_optimization_pass(
+    run_after: Iterable[str] = tuple(),
```
Do we expect users to pass in run after names or is this mainly internal configuration (except for expert users who want to customize graph optimization)

fegin replied:
This is only for developers who develop graph optimization passes to specify the order constraint.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+                gm, (fx.GraphModule, IterGraphModule)
+            ), "The first argument of the pass must be either fx.GraphModule or IterGraphModule"
+            assert (
+                func.__name__ not in _optimized_func
```
this seems to be brittle, as functions with the same name would collide even if they are from different packages?

fegin replied:
The very original design is to put all the optimization passes in a class. This way, the name will guarantee to be unique. But I feel it is a little bit overkill to use another class. Any suggestion of this?

fegin replied:
On second thought, the design should be fine.A pass will only be added to `_optimized_func` if the pass is wrapped with `graph_optimization_pass` and run. If there are more than one optimization passes that have the same naming, I feel this is actually good to error out because the same naming will eventually cause problems.

mrshenli replied:
> Any suggestion of this?

Maybe this?

```python
".".join([func.__module__, func.__name__])
```

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+
+def get_comm_block_nodes(comm_node: fx.Node) -> CommBlock:
+    """
+    Given a wait_comm node, find out all the nodes belong to this communcation.
```
nit: with AsyncTensor, this is now `aten.wait_tensor.default`

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+    wait_nodes = []
+    inputs, _ = tree_flatten((comm_node.args, comm_node.kwargs))
+    distance = 0
+    wait_prefixes = ("wait_comm", "wait_tensor")
```
do we still need to support `wait_comm` from `CommTensor`?

fegin replied:
Nope, this can be removed.

fegin replied:
Wait, we may still need to support `wait_comm`. Alltoall is still using `CommTensor`?

mrshenli replied:
> Alltoall is still using CommTensor?

Ah, yep for now. Need @kumpera 's help to get A2A to functional collectives :)

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+    non_end_users_nodes = ("split", "reshape", "getitem", "detach", "alias")
+
+    nodes = collections.deque([comm_node, None])
+    while nodes and distance < 5:
```
any reason we pick 5 as the distance here?

mrshenli replied:
Oh I see, if this is the first pass after getting DTensor expanded graph, the wait should be right after allreduce as the latter is triggered on demand.

mrshenli replied:
But does this still hold for sparse? cc @yifuwang

fegin replied:
5 is just a safety guard based again on the expansion as you mentioned. However, even if there are optimization pass that move wait around, the distance should not change? The BFS directly traverse the users. So theoretically, the distance between comm_ops and wait_tensor is 1. 5 is used for wait_comm

mrshenli replied:
Ah, yep, you are right, this is BFS distance

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+    while nodes:
+        node = nodes.popleft()
+        for user in node.users:
+            if isinstance(user, fx.Node) and user.name.startswith(non_end_users_nodes):
```
curious, when can a user of fx.Node be non-fx.Node?

fegin replied:
Well, we added `__hold__` for the users of some nodes to avoid DCE in the old tracing scheme. This can probably  be removed as we no longer need to do so.

mrshenli replied:
oh I see!

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+                nodes.append(user)
+                node_list.append(user)
+            else:
+                outputs[node] = None
```
this always maps to None?

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+    inputs, _ = tree_flatten((comm_node.args, comm_node.kwargs))
+    distance = 0
+    wait_prefixes = ("wait_comm", "wait_tensor")
+    non_end_users_nodes = ("split", "reshape", "getitem", "detach", "alias")
```
these means non-compute node? what about view and transpose?

fegin replied:
These nodes are added by our tracing and graph optimizations. But we can add view and transpose as well.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+        )
+
+    # Identify all the outputs
+    outputs: Dict[fx.Node, Any] = {}
```
curious, why do we call this `outputs`? It feel more like value consumer nodes?

mrshenli replied:
oh, nvm, they are outputs for the comm block

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+            last_input_node = input_node
+            last_input_index = index
+
+    with gm.graph.inserting_after(last_input_node):
```
curious, will other graph optimization passes move any of these input compute nodes around?

fegin replied:
Right now, no. Are you suggesting that we should move the input nodes earlier?

mrshenli replied:
I do not know. What if we fond an independent compute block (which contains the input) that we wanna move to an earlier place to overlap with other comms? Just a random thought.

fegin replied:
That's actually very possible I think. We should try if to squeeze more from the computation graph. Right now we just change the communication order.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+    for node in gm.graph.nodes:
+        if node == fused_comm_block.comm_node:
+            break
+        last_wait_node_idx = max(
```
just for discussion, we rely on the topological order given by fx.Graph, but do we need worry about that order is not a preferred topological order for us? E.g., a different topo order might allow us to delay wait a bit more? Will there be additional passes to further optimize that?

fegin replied:
Yes, as we discussed offline, the graph which generated by shared weights will not be optimized well without changing the topological sort order. I have another implementation which move the compute nodes between comm_ops and wait instead of moving wait. I feel that one can solve this issue, but that implementation still needs to be polished.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,408 @@
+    )
+
+
+def _call_function(
```
it seems might be possible to consolidate this with the sym_call family from @yifuwang's diff :)

fegin replied:
Not aware of that one, but will check and we should definitely consolidate the calls.

fegin committed:
Update on "[SPMD] Introduce graph_optimization_pass and comm_fusion_with_cat"


This PR add `graph_optimization_pass` decorator which should be wrapped by all graph optimization passes. This PR also introduces the first graph optimization, `comm_fusion_with_cat`, as the first use case of `graph_optimization_pass`.

Differential Revision: [D44661608](https://our.internmc.facebook.com/intern/diff/D44661608/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce graph_optimization_pass and comm_fusion_with_cat"


This PR add `graph_optimization_pass` decorator which should be wrapped by all graph optimization passes. This PR also introduces the first graph optimization, `comm_fusion_with_cat`, as the first use case of `graph_optimization_pass`.

Differential Revision: [D44661608](https://our.internmc.facebook.com/intern/diff/D44661608/)

[ghstack-poisoned]

yifuwang commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,440 @@
+
+    tensor_meta = cat_node.meta.get("tensor_meta")
+    fused_comm_block = CommBlock(
+        shape=tensor_meta.shape,  # type: [union-attr]
```
When the input graph has symbolic shape enabled, `.shape` will contain `SymInt`s and `comm_fusion_with_cat` will break the graph by using `SymInt`s not defined in the graph as some nodes' arguments.

In the long term we need to support symbolic shape. For now we need to convert the `SymInt`s to `int`s here (e.g. `shape=torch.Size(int(s) for s in torch_meta.shape)`).

fegin replied:
Applying the change and will add a TODO here.

yifuwang commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -0,0 +1,440 @@
+            aten.split,
+            (
+                fused_wait_node,
+                [cast(torch.Size, cb.shape).numel() for cb in comm_blocks],
```
`.numel()` can be a `SymInt`. We need to converted it to `int`.

fegin replied:
Thanks for catching the error. Applying the changes to the PR.

fegin committed:
Update on "[SPMD] Introduce graph_optimization_pass and comm_fusion_with_cat"


This PR add `graph_optimization_pass` decorator which should be wrapped by all graph optimization passes. This PR also introduces the first graph optimization, `comm_fusion_with_cat`, as the first use case of `graph_optimization_pass`.

Differential Revision: [D44661608](https://our.internmc.facebook.com/intern/diff/D44661608/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce graph_optimization_pass and comm_fusion_with_cat"


This PR add `graph_optimization_pass` decorator which should be wrapped by all graph optimization passes. This PR also introduces the first graph optimization, `comm_fusion_with_cat`, as the first use case of `graph_optimization_pass`.

Differential Revision: [D44661608](https://our.internmc.facebook.com/intern/diff/D44661608/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce graph_optimization_pass and comm_fusion_with_cat"


This PR add `graph_optimization_pass` decorator which should be wrapped by all graph optimization passes. This PR also introduces the first graph optimization, `comm_fusion_with_cat`, as the first use case of `graph_optimization_pass`.

Differential Revision: [D44661608](https://our.internmc.facebook.com/intern/diff/D44661608/)

[ghstack-poisoned]
